runtime/devnet/src/extensions/mod.rs

```diff
@@ -0,0 +1,1160 @@
+}
+
+fn dispatch_call<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	call: RuntimeCall,
+	mut origin: RuntimeOrigin,
+	log_prefix: &str,
+) -> Result<(), DispatchError>
+where
+	T: frame_system::Config<RuntimeOrigin = RuntimeOrigin, RuntimeCall = RuntimeCall>,
+	RuntimeOrigin: From<RawOrigin<T::AccountId>>,
+	E: Ext<T = T>,
+{
+	let charged_dispatch_weight = env.charge_weight(call.get_dispatch_info().weight)?;
+	log::debug!(target:LOG_TARGET, "{} Inputted RuntimeCall: {:?}", log_prefix, call);
+	origin.add_filter(AllowedApiCalls::contains);
+	match call.dispatch(origin) {
+		Ok(info) => {
+			log::debug!(target:LOG_TARGET, "{} success, actual weight: {:?}", log_prefix, info.actual_weight);
+			// Refund weight if the actual weight is less than the charged weight.
+			if let Some(actual_weight) = info.actual_weight {
+				env.adjust_weight(charged_dispatch_weight, actual_weight);
+			}
+			Ok(())
+		},
+		Err(err) => {
+			log::debug!(target:LOG_TARGET, "{} failed: error: {:?}", log_prefix, err.error);
+			Err(err.error)
+		},
+	}
+}
+
+fn construct_call(
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<RuntimeCall, DispatchError> {
+	match pallet_index {
+		52 => {
+			let call = versioned_construct_assets_call(version, call_index, params)?;
+			Ok(RuntimeCall::Assets(call))
+		},
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn construct_key(
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<RuntimeStateKeys, DispatchError> {
+	match pallet_index {
+		52 => {
+			let key = versioned_construct_assets_key(version, call_index, params)?;
+			Ok(RuntimeStateKeys::Assets(key))
+		},
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn versioned_construct_assets_call(
+	version: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<pallet_assets::Call<Runtime, TrustBackedAssetsInstance>, DispatchError> {
+	match version {
+		V0 => v0::assets::construct_assets_call(call_index, params),
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn versioned_construct_assets_key(
+	version: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<AssetsKeys, DispatchError> {
+	match version {
+		V0 => v0::assets::construct_assets_key(call_index, params),
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn read_state<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<(), DispatchError>
+where
+	T: pallet_contracts::Config
+		+ pallet_assets::Config<TrustBackedAssetsInstance, AssetId = AssetId>
+		+ pallet_nfts::Config<CollectionId = CollectionId, ItemId = ItemId>
+		+ cumulus_pallet_parachain_system::Config
+		+ frame_system::Config<AccountId = sp_runtime::AccountId32>,
+	E: Ext<T = T>,
+{
+	const LOG_PREFIX: &str = " read_state |";
+	let key = construct_key(version, pallet_index, call_index, params)?;
```

Comment:
As above, decoding into an enum would be much easier.

Reply:
Included it in issue description #94